### ISSUE TITLE
Use memory bank for images not belonging to the same video ?

### ISSUE THREAD

**Issue by tcourat** [facebookresearch/sam2#352]
Has anyone tried to leverage the SAMv2 memory bank for other purpose than video segmentation ? For instance if I have 10 images of a given object (like a table, chair etc..), I only click on the object  in image 1 and want sam to automatically segment similar objects in all the other images ?

**Comment by heyoeyo** [comment]
There was a related discussion in issue #210, though that was for using an image as a starting prompt for a video. In general it can work, but it's probably not capable of exact matching (i.e. I wouldn't expect it to pick out a specific person from a crowd, though I haven't tried). It seems to match based on multiple factors, not just the 'semantic' label we would assign but also color and positioning (of the original prompt) for example.

Here's a simple example of prompting on a picture of a bird and applying that to another similar picture along with an unrelated pictures (these photos are all from the [BIG](https://github.com/hkchengrex/CascadePSP/blob/master/docs/dataset.md) dataset):

Prompt image:
![Screenshot from 2024-10-03 09-49-03](https://github.com/user-attachments/assets/9ce9ddb7-54bc-4fdf-b4f8-d7c282aaef92)

Transfer image (i.e. no prompt used, instead it uses the memory bank from the result above). It may prefer the central bird because the original prompt is centered, even though the left bird seems like a better match.
![Screenshot from 2024-10-03 09-49-34](https://github.com/user-attachments/assets/75a6e3bd-037d-46e1-8058-0a777ef6ad26)

Two more transfer images, but unrelated to the original. Here it is picking stuff on the left, maybe because of the sizing of the objects (not really sure).
![Screenshot from 2024-10-03 09-49-08](https://github.com/user-attachments/assets/29ded62a-31f3-4a74-aa61-f4194ae1f824)
![Screenshot from 2024-10-03 10-04-28](https://github.com/user-attachments/assets/e7fff8e5-be15-4a32-a63a-4ac839905c9f)

**Comment by heyoeyo** [comment]
Just as a follow-up for anyone interested, I've set up an [interactive script](https://github.com/heyoeyo/muggled_sam/tree/main/experiments#cross-image-segmentation) to make this easier to try. Here's a quick example video:

https://github.com/user-attachments/assets/b9459548-3b94-4106-ad3e-86fc794349e9

On the left is an image being prompted with a single foreground point (the green dot), while the image on the right is segmented based on the 'video memory bank' from the first image result. The model seems to intutively segment things like the eye, beak and feet as well as the branch/platform the birds stand on. It gives lots of weird and often bad results too, and is very sensitive to which mask is being selected.

For anyone trying to use the model this way, it seems important to encode the second image in the memory bank several times (i.e. as if it appears repeatedly as a short video) for the segmentation to come out cleanly.

**Comment by mgrewe** [comment]
That is interesting. Do you mean that the segmentation on the right becomes better the more often the segmentation is repeated? Would you suspect this to be some kind of averaging/smoothing effect? Would be glad if you could share some example images.

**Comment by heyoeyo** [comment]
> Would you suspect this to be some kind of averaging/smoothing effect?

It does frequently have a smoothing effect. It often seems to converge on a better mask of the area that was initially (uncleanly) masked. This clean up trick also works when cross-prompting an image with itself (though maybe simpler to just adjust the prompt).

Here's an example where the initial (bird) prompt causes both sheep to get messy masks, but repeating the frames cleans up the mask on the left sheep and removes the right sheep:

https://github.com/user-attachments/assets/0944317f-4442-478c-b1bb-417cb8e23b5c

Interestingly, it also improves the object score which is normally an indicator for the video tracking that it 'found' the object in the next frame. It also converges on the legs of the sheep in the second mask output (far right, second from the top) and sort of converges on the sheep's face on the last mask output, so the behavior varies based on the initial mask.

**Comment by mgrewe** [comment]
Thanks @heyoeyo . Very interesting insight!

**Comment by heyoeyo** [comment]
Unfortunately, the SAMv1 models can't be used in the same way since they don't have the video capability/memory bank of the v2 models.

**Comment by mgrewe** [comment]
Yeah, I know. Rather meant only repeatedly feeding in a mask to improve it. Was hoping to use it for improvement of solutions from using a mask as input prompt - which to my knowledge gives worse results as compared to other prompt types. But that is slightly off topic here.

**Comment by heyoeyo** [comment]
> Rather meant only repeatedly feeding in a mask to improve it

Oh I see what you mean. I haven't experimented with the mask prompts too much, since (as you say) they tend to give disappointing results compared to other prompts. From what little I have seen, the mask prompts (v1 or v2) don't seem to show the same convergence behavior as using the v2 memory bank. Surprisingly, the v2.1 mask prompts do show some cross-image prompting capability though (could be a coincidence, I haven't tried it on many images).

That being said, in the [SAMv1 paper](https://arxiv.org/pdf/2304.02643) they mention (on page 17 under the **Training Algorithm** section) feeding the output of the model back in as a mask prompt as if it helps, so maybe there is a way to make it work. The automatic mask generator even includes this [sort of processing](https://github.com/facebookresearch/sam2/blob/c98aa6bea377d5c000cdc80197ce402dbf5304dc/sam2/automatic_mask_generator.py#L344) as a option.
